### DISCUSSION TITLE
Is there a way to pre load state on app start (before material app runs)

### DISCUSSION THREAD

**Discussion by danieldoddjr** [rrousselGit/riverpod#842] [Q&A]
**Describe what scenario you think is uncovered by the existing examples/articles**
I am dynamically building routes (vRouter) using a change notifier waiting on an API call. Is there a way to preload state before material app is built so I can make sure my routes are ready? 

**Describe why existing examples/articles do not cover this case**
I have not seen anything that covers pre loading state before building the app in the documentation.

**Comment by rrousselGit** [comment]
You can read your providers at any time, which will cause them to load.

**Comment by esDotDev** [comment]
It could just be something like:
```
_MainViewState extends ConsumerState {
   bool _readyToBuildRoutes = false;

  void initState(){
     ref.read(apiProvider).call().then((_) => setState(() => _readyToBuildRoutes = true));
  }
}
```

**Comment by hpoul** [comment]
fwiw, I am loading user configuration from disk before calling `runApp`.. (I think this way the splash screen is shown until i'm actually ready). The way I did it was basically to have a `final appDataProvider = Provider<AppData>((ref) => throw Exception());`

and then overwrite the value in the `ProviderScope` - not sure if this is an anti pattern though 🤷‍♂️ especially if you do network requests instead of just loading from disk you might require a progress/error indicator anyway..

```dart
  Future<void> main() async {
    final appData = await loadData();
    runApp(ProviderScope(
      overrides: [
        appDataProvider.overrideWithValue(appData),
      ],
      child: App(),
    );
  }
```

**Comment by ghost** [comment]
I'm using splash screen with value notifier to initialize services/notifiers and then override providers. It's probably not the cleanest approach but gets the job done. A good example of such approach is in this repo [memo](https://github.com/olmps/memo)

**Reply by hpoul** [reply]
This seems to just implement a separate splash screen in flutter.. imho it's preferable when the native splash screen keeps showing until initial loading is done? 🤔 
I think this can only be done by initializing before calling `runApp`? 

> Flutter automatically keeps the Android launch screen displayed until Flutter has drawn the first frame

**Reply by ghost** [reply]
tbh I'm not sure which approach is better, but I guess if initialization is going to take more than 200-300ms you'll need to show some custom widgets(that might not be easy to implement for multiple platforms since each have slightly different initial screen). you can use runApp more than once but that might have some problems as well
there are good discussions [here ](https://github.com/rrousselGit/river_pod/issues/329)
